apps/explorer/lib/explorer/chain/optimism/withdrawal.ex

```diff
@@ -269,33 +286,83 @@ defmodule Explorer.Chain.Optimism.Withdrawal do
     end
   end
 
-  defp handle_proven_status({l1_timestamp, game_index}, respected_games) do
-    game = game_by_index(game_index)
-
-    cond do
-      is_nil(game_index) and not Enum.empty?(respected_games) ->
-        # here we cannot exactly determine the status `Waiting a game to resolve` or
-        # `Ready for relay` or `In challenge period`
-        # as we don't know the game index. In this case we display the `Proven` status
-        {@withdrawal_status_proven, nil}
-
-      is_nil(game) or DateTime.compare(l1_timestamp, game.created_at) == :lt ->
-        # the old status determining approach
-        pre_fault_proofs_status(l1_timestamp)
-
-      true ->
-        # the new status determining approach
-        post_fault_proofs_status(l1_timestamp, game)
-    end
+  # Determines the current withdrawal status by the list of the bound WithdrawalProven events.
+  #
+  # ## Parameters
+  # - `proven_events`: A list of WithdrawalProven events. Each item is `{l1_timestamp, game_index}` tuple.
+  # - `respected_games`: A list of games returned by the `respected_games()` function.
+  #
+  # ## Returns
+  # - `{status, datetime}` tuple where the `status` is the current withdrawal status,
+  #                        `datetime` is the point of time when the challenge period ends.
+  #                        (only for `In challenge period` status).
+  @spec handle_proven_status(list(), list()) :: {String.t(), DateTime.t() | nil}
+  defp handle_proven_status(proven_events, respected_games) do
+    statuses =
+      proven_events
+      |> Enum.reduce_while([], fn {l1_timestamp, game_index}, acc ->
+        game = game_by_index(game_index)
+
+        status =
+          cond do
+            is_nil(game_index) and not Enum.empty?(respected_games) ->
+              # here we cannot exactly determine the status `Waiting a game to resolve` or
+              # `Ready for relay` or `In challenge period`
+              # as we don't know the game index. In this case we display the `Proven` status
+              {@withdrawal_status_proven, nil}
+
+            is_nil(game) or DateTime.compare(l1_timestamp, game.created_at) == :lt ->
+              # the old status determining approach
+              pre_fault_proofs_status(l1_timestamp)
+
+            true ->
+              # the new status determining approach
+              post_fault_proofs_status(l1_timestamp, game)
+          end
+
+        if elem(status, 0) == @withdrawal_status_ready_for_relay do
+          {:halt, [status]}
+        else
+          {:cont, [status | acc]}
+        end
```

Comment:
Maybe we can pattern match on `status`?

```suggestion
        cond do
          is_nil(game_index) and not Enum.empty?(respected_games) ->
            # here we cannot exactly determine the status `Waiting a game to resolve` or
            # `Ready for relay` or `In challenge period`
            # as we don't know the game index. In this case we display the `Proven` status
            {@withdrawal_status_proven, nil}

          is_nil(game) or DateTime.compare(l1_timestamp, game.created_at) == :lt ->
            # the old status determining approach
            pre_fault_proofs_status(l1_timestamp)

          true ->
            # the new status determining approach
            post_fault_proofs_status(l1_timestamp, game)
        end
        |> case do
          {@withdrawal_status_ready_for_relay, _} -> {:halt, [status]}
          _ -> {:halt, [status]}
        end
```

Reply:
Ok, but I had to add the comment for credo: https://github.com/blockscout/blockscout/pull/11399/commits/a7307c8dc754873b01cfcdc171c6d72a14ffbcb1